feat(ui): auto-switch inactive headers to icons when width is limited

glmlm opened the pull request:
See #87 

I'm not entirely confident that the implementation or variable naming is optimal yet, so I'd appreciate your feedback on those points.
Happy to revise based on your suggestions.

glmlm committed:
feat(ui): auto-switch inactive headers to icons when width is limited

igorlfs commented on lua/dap-view/config.lua:
```diff
@@ -46,6 +46,7 @@ local M = {}
 
 ---@class dapview.SectionConfig
 ---@field label string
+---@field label_min string
```
```suggestion
---@field short_label string
```

igorlfs commented on lua/dap-view/options/winbar.lua:
```diff
@@ -45,7 +45,11 @@ local set_winbar_opt = function()
             if section ~= nil then
-                local desc = " " .. section.label .. " "
+                local is_current_section = state.current_section == v
+                local width = vim.api.nvim_win_get_width(state.winnr)
```
As a convention, I like to abbreviate `vim.api` as `api`, with a local on module level (check other files).

igorlfs commented on lua/dap-view/options/winbar.lua:
```diff
@@ -45,7 +45,11 @@ local set_winbar_opt = function()
-                local desc = " " .. section.label .. " "
+                local is_current_section = state.current_section == v
+                local width = vim.api.nvim_win_get_width(state.winnr)
+                local width_limit = winbar.controls.enabled and 105 or 81
```
We can't hardcode these values, as the sections are customizable (e.g., someone could be using a single section). We would have to calculate the length of each enabled section and then check against this value. The same also goes for the control bar, someone could be using a subset of the buttons.

igorlfs replied:
Alternatively, we could add a switch to toggle between the short / long version.

glmlm replied:
I've updated the logic to dynamically calculate the length based on the enabled sections.
Note that `controls_len` currently assumes each item is a single character, so I can adjust that as well if necessary.

igorlfs commented on lua/dap-view/options/winbar.lua:
```diff
@@ -45,7 +45,11 @@ local set_winbar_opt = function()
+                local is_current_section = state.current_section == v
+                local width = vim.api.nvim_win_get_width(state.winnr)
+                local width_limit = winbar.controls.enabled and 105 or 81
+                local label = not is_current_section and width < width_limit and section.label_min or section.label
```
I wonder if it also makes sense to shorten the current section?

igorlfs reviewed (commented):
I'm afraid this issue won't be as straightforward as anticipated 🤔

glmlm committed:
refactor(ui): improve header shortening logic

- Rename `label_min` to `short_label`
- Introduce local `api` alias for `vim.api`
- Make the threshold dynamic based on sections and controls

glmlm committed:
fix(ui): account for double-width characters in label width calculation

glmlm committed:
fix(ui): incorrect calculation of `controls_len`

igorlfs commented on lua/dap-view/options/winbar.lua:
```diff
@@ -45,7 +47,40 @@ local set_winbar_opt = function()
+                    vim.tbl_values(winbar.custom_sections),
+                })
+                    :flatten()
+                    :map(function(sec)
```
Can you add some type annotations for these anonymous functions?

igorlfs replied:
Looks alright! Made me notice I usually don't add the return type if it's not ambiguous, but that might be a good idea

igorlfs commented on lua/dap-view/options/winbar.lua:
```diff
@@ -45,7 +47,40 @@ local set_winbar_opt = function()
+                    :fold(0, function(len_total, len_label)
+                        return len_total + len_label
+                    end)
+                local width_limit = winbar.controls.enabled and labels_len + controls_len or labels_len
```
We should cache the length to avoid calculating the same thing over and over again. Changing the winbar's components while the plugin is running is a niche use case.

glmlm replied:
I added caching by storing the values in local variables outside the function, initializing them only if they haven’t been set yet.
Does this approach seem appropriate to you, or would you prefer a different one?

igorlfs replied:
> Does this approach seem appropriate to you, or would you prefer a different one?

That was exactly what I was thinking!

glmlm committed:
refactor: extract section length calculation with caching and type annotations

- Extract calculation logic of `labels_len` and `controls_len` into separate functions for clarity
- Add caching for `labels_len` and `controls_len` to avoid redundant calculations
- Add type annotations for anonymous functions

igorlfs commented:
I'm gonna merge as is, and add some refactors afterwards

igorlfs commented:
> I'm gonna merge as is, and add some refactors afterwards

I shouldn't have rushed to merge this PR. The logic to only handle the enabled sections / buttons was incomplete. But after investigating the situation, I came to the conclusion that handling this logic for the buttons would be a bit nasty. 

So I decided to completely rewrite the buttons, similarly to the sections.

I also made some other simplifications. Check the [commit](https://github.com/igorlfs/nvim-dap-view/pull/85/commits/462f0132e7d84e8a0195ed7759daa87da78a641e) if you wanna learn more.

By the way, can you give #85 a try and check if everything is working?

igorlfs commented:
Hey @glmlm, sorry for the repeated ping, I was revisiting this problem and I noticed I could have documented this transition a little better.

One can define a "maximum length" (by hand) and then use something like:

```lua
return {
    winbar = {
        base_sections = {
            watches = { label = function(width) return width < 120 and "Watches" or "󰛐 " end },
            -- ...
        }
    }
}
```
